### ISSUE TITLE
Running model prediction on samples of unknown batch size

### ISSUE THREAD

**Issue by guileryu01** [dmlc/mxnet#2691]
It was my understanding that when calling simple_bind to bind a symbol to an executor one needs to provide input shape. For convolutional net input shape often takes the form of (batch_size, n_feature, width, height). It is fine for training purpose to fix batch size upfront, but if I want to build an application that takes a pretrained model the run prediction with it, the number of samples for each call is typically unknown until run time. Is there any suggestion on how to handle this efficiently (suppose evaluating these samples is performance critical part within the whole process)?

**Comment by piiswrong** [comment]
try mx.module.bucketingmodel

**Comment by guileryu01** [comment]
Thanks very much for the quick reply.
I didn't know about mx.module. By reading through the code, it looks like BucketingModule allows one to use multiple bucket key, each of which is associated with an executor. Does it mean that if I want to use it for prediction on a generic DataBatch, I would need to create one executor per batch size (and copy weight matrices to each of them)?

**Comment by piiswrong** [comment]
Weight matrices are shared. So are internel buffers

**Comment by guileryu01** [comment]
Thank you, I will give that a try.
On the other hand, I was wondering if the requirement to binding a symbol can be relaxed a bit so that it allows an executor to take inputs with shapes that are only different on their first dimension (batch size)? I feel that this is natural as most of the execution is shared after all.

**Comment by piiswrong** [comment]
It doesn't matter which dim is different. The memory buffer size will be different in the end. That's why you need to bind again.
